[Security] Extract password hashing from security-core - with proper wording

chalasr opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fixes #39698
| License       | MIT
| Doc PR        | todo

This PR renames password "encoders" to password _hashers_ (naming widely used, see e.g. django or laravel).
This also takes the opportunity to extract the logic related to password hashing from security-core, moving it to a new password-hasher component.
Nowadays, many modern web apps and APIs don't deal with passwords at all, that's why splitting makes sense as a step towards making security-core not tied to the password concept.

For upgrading, applications will have to use `passwords_hashers` instead of `encoders` in their security configuration,  and type-hint against `PasswordHasherInterface` (and related) instead of `PasswordEncoderInterface`.

The proposed API is not much different from the encoder one regarding behavior and signatures, and it is slightly more close to the PHP built-in password hashing API:

```php
namespace Symfony\Component\PasswordHasher;

interface PasswordHasherInterface
{
    public function hash(string $plainPassword): string;

    public function verify(string $hashedPassword, string $plainPassword): bool;

    public function needsRehash(string $hashedPassword): bool;
}
```

chalasr committed:
[Security] Extract password hashing from security-core - using the right naming

stof commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php:
```diff
@@ -355,6 +356,7 @@ private function addEncodersSection(ArrayNodeDefinition $rootNode)
             ->fixXmlConfig('encoder')
             ->children()
                 ->arrayNode('encoders')
+                ->setDeprecated('symfony/security-bundle', '5.3', 'The child node "%node%" at path "%path%" is deprecated, use "password_hashers" instead.')
```
should be indented

nicolas-grekas commented on src/Symfony/Component/Security/Password/SaltingPasswordHasherInterface.php:
```diff
@@ -0,0 +1,38 @@
+ * @author Nicolas Grekas <p@tchwork.com>
+ * @author Robin Chalas <robin.chalas@gmail.com>
+ */
+interface SaltingPasswordHasherInterface extends PasswordHasherInterface
```
why not `SelfSaltingPasswordHasherInterface`?

Also, would there be a way to remove this interface and merge it with `PasswordHasherInterface`? Aka remove the concept of salt from our interfaces - since it's an obsolete one?

chalasr replied:
This one does the inverse of the current `SelfSaltingEncoderInterface`: it should only be implemented by implementations that need user-provided salts, so that modern hashers (native, sodium etc) don't take a no-op `salt` argument.
Removing the concept entirely would be hard since we still have old hashers that require a salt... so the idea is to have this interface until we remove those old hashers

stof replied:
but `SaltingPasswordHasherInterface` is a bad name for such hasher. Modern hashers are also salting the password, but they do that without requiring an external salt.

nicolas-grekas replied:
`LegacyPasswordHasherInterface` to make this clear?

chalasr replied:
👍 thanks, renamed

stof commented on src/Symfony/Component/Security/Core/Encoder/UserPasswordEncoder.php:
```diff
@@ -11,12 +11,17 @@
 
 namespace Symfony\Component\Security\Core\Encoder;
 
+trigger_deprecation('symfony/security-core', '5.3', sprintf('The "%s" interface is deprecated, use "%s" on hasher implementations that deal with salts instead.', UserPasswordEncoder::class, UserPasswordHasher::class));
```
wrong message

stof commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php:
```diff
@@ -168,13 +172,31 @@ public function load(array $configs, ContainerBuilder $container)
+        // @deprecated since Symfony 5.3
         if ($config['encoders']) {
             $this->createEncoders($config['encoders'], $container);
+            $config['password_hashers'] = $config['encoders'];
```
this assignment means that if you explicitly configure both, the new non-deprecated config is silently lost. This looks wrong to me.

stof replied:
if we want to consider both configs equivalent (as suggested below), my suggestion would actually be perform this rewriting on a per-file basis in the Configuration class (using `beforeNormalization()`, not `validate()`) so that the usage of the old API in one of the file does not make you loose the config done in the modern way in the other file.

stof commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php:
```diff
@@ -168,13 +172,31 @@ public function load(array $configs, ContainerBuilder $container)
+        }
+
+        if ($config['password_hashers']) {
+            $this->createHashers($config['password_hashers'], $container);
```
I see an issue here: if a project migrate their config to use `password_hashers`, the corresponding encoder services are not defined anymore. And this means that any bundle which is not migrated yet to the new API will break as `security.encoder_factory` won't be able to give it an encoder.
As the feature set is the same, I think we should create legacy services as well when using the new config (similar to how the old config creates new services with your assignment above). If you never use the old `security.encoder_factory` API, having these deprecated services defined won't hurt (deprecated services are allowed to use deprecated classes without triggering a compile-time deprecation warning for that in the compiler pass IIRC) and it will keep BC for services using them (with a deprecation warning)

stof replied:
btw, this also implies that both the new and the old factories should support both `*Aware` interfaces implemented on User classes, so that both factories can be used for migrated and non-migrated user classes while getting the right hasher/encoder.

derrabus commented on src/Symfony/Component/Security/Core/Encoder/EncoderAwareInterface.php:
```diff
@@ -1,5 +1,6 @@
 <?php
 
+
```
Should be removed 🙃

derrabus commented on src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php:
```diff
@@ -120,8 +127,13 @@ private function authenticateViaGuard(AuthenticatorInterface $guardAuthenticator
-        if ($this->userProvider instanceof PasswordUpgraderInterface && $guardAuthenticator instanceof PasswordAuthenticatedInterface && null !== $this->passwordEncoder && (null !== $password = $guardAuthenticator->getPassword($token->getCredentials())) && method_exists($this->passwordEncoder, 'needsRehash') && $this->passwordEncoder->needsRehash($user)) {
-            $this->userProvider->upgradePassword($user, $this->passwordEncoder->encodePassword($user, $password));
+        if ($this->userProvider instanceof PasswordUpgraderInterface && $guardAuthenticator instanceof PasswordAuthenticatedInterface && null !== $this->passwordHasher && (null !== $password = $guardAuthenticator->getPassword($token->getCredentials())) && method_exists($this->passwordHasher, 'needsRehash') && $this->passwordHasher->needsRehash($user)) {
+            if ($this->passwordHasher instanceof PasswordEncoderInterface) {
+                // @deprecated since Symfony 5.3
+                $this->userProvider->upgradePassword($user, $this->passwordHasher->encodePassword($user, $password));
+            } else {
+                $this->userProvider->upgradePassword($user, $this->passwordHasher->encodePassword($user, $password));
+            }
```
`if` and `else` branch execute the same logic. You probably wanted to call the new method in the `else` branch.

chalasr replied:
nice catch! fixed

stof commented on src/Symfony/Component/Security/Password/Hasher/HasherFactoryInterface.php:
```diff
@@ -0,0 +1,32 @@
+/**
+ * HasherFactoryInterface to support different hashers for different accounts.
+ *
+ * @author Johannes M. Schmitt <schmittjoh@gmail.com>
```
not really true for the new interface

derrabus commented:
I like the idea of moving password hasing out of security-core. Would it make sense to create a contracts package for the new hasher interfaces?

stof commented:
I don't think these interfaces need to be in contracts. HasherInterface is not something you would typehint anyway most of the time (you need the factory to get the right hasher for that user class).
And HasherFactoryInterface is not really a contract that make sense to reimplement outside `symfony/security-*` as it depends on other concepts there. So to me, there is no reason to use the interfaces without also using the component itself.

goetas commented:
A note on the interface implementation, it happened multiple times that the neeed for a re hashing of a password was something depended on the user/token holding it, not just the password itself. Would it be possible to talke that into consideration somehow?

apfelbox commented:
> I don't think these interfaces need to be in contracts. HasherInterface is not something you would typehint anyway most of the time (you need the factory to get the right hasher for that user class).

But you could use it in your library that provides a new hasher? And then you would ideally only need to implement the interface from the contract, instead of tying it to the whole package?

But OTOH it's simple enough for your bundle to just allow "both" versions (in case of a BC), so... probably not worth the effort.

chalasr commented:
> A note on the interface implementation, it happened multiple times that the neeed for a re hashing of a password was something depended on the user/token holding it, not just the password itself. Would it be possible to talke that into consideration somehow?

@goetas If you are talking about the way built-in opportunistic password migrations happen, yes, PasswordUpgraderInterface requires a user in order to work, and it's is usually implemented by UserProviders.
However, `PasswordHasherInterface::needsRehash()` does not know about the user, nor `MigratingPasswordHasher` does: hashing options are set on the implementation (usually via constructor). As such, changing the values of those options (or using a different implementation) should make `needsRehash()` return true, so you can handle rehashing on your own.

> But you could use it in your library that provides a new hasher? And then you would ideally only need to implement the interface from the contract, instead of tying it to the whole package?

Yes, I can imagine this component being used as a simple hashing library outside of the framework, in which case one could perfectly rely on this interface rather than the factory. Might be useful for interop also.
I'll let others decide on this one :) 

@stof Thanks for the great review. I addressed your comments.

This is ready for review.

stof commented:
> Yes, I can imagine this component being used as a simple hashing library outside of the framework, in which case one could perfectly rely on this interface rather than the factory. Might be useful for interop also.

but if you use the component as hashing library, you need the component implementation anyway, not only the interface.

apfelbox commented:
> but if you use the component as hashing library, you need the component implementation anyway, not only the interface.

As a **user** of the hashing library: yes

However as somebody, who just wants to write a package that provides eg. three hashing algorithms for usage with this library: nope 😉

wouterj commented:
Thanks @chalasr, this is much more than I could have imagined when you said you would do the renaming. I like it! ;-)

I've had a quick look and I have 2 global questions about these changes:

- I believe the security sub-components are a mistake, I would prefer if we make this a root level component instead (see #36423 for some background on this);
- I think hashers are useful in a broader area than just user passwords, and I feel like splitting this logic in a separate component opens up our hashers for this broader area. I think the name is a bit unfortunate here and maybe "Hash" is better than "Password". 
  <s>I think we can even consider merging this effort with @Nyholm's PR introducing the Encryption component (ref #39344) and call it e.g. the "Cryptography" component, providing encryption and hashing.</s>

What do you think?

stof commented:
@wouterj given that hashing and encrypting have totally different use cases (that should not be mixed), I suggest to keep them separate. We don't want to reuse the same interfaces for those (that would not be possible anyway AFAICT as the possibilities are different), so putting them in the same component would be similar to creating 2 subcomponents in `symfony/cryptography`

wouterj commented:
Alright, makes sense. It makes some sense as a "grouping" method, but you indeed wouldn't ever want to switch the implementations.

stof commented:
@wouterj I would even vote for keeping `password` in the name. This component is meant to provide password hashing features, which are only a subset of crypto hashing. Not all crypto hashing algorithms are suitable for password hashing (as they need to protect against brute force attacks and so they need to be slow)

derrabus commented:
Thank you very much for this PR, @chalasr! I see @wouterj's point that we shouldn't create new Security sub-components and agree with @stof that we shouldn't mix the concern of password hashing with encryption. A top-level PasswordHasher oder PasswordHashing component would be my personal preference here.

wouterj commented:
Hmm, if this component is so specific that it can only work with user passwords, I'm unsure if it makes sense to split it out of the Security Core component. From a usage point of view, it means developers have to do an extra step when using passwords (install a package), which we also have to document. I feel like, although APIs are not protected by passwords, almost any online service is still password protected in some way (even Slack has passwords next to their magic links).

The goal of making Security less dependent on passwords is a different topic from extracting it imho. E.g. the `getPassword()` method in `UserInterface` and the requirement to specify a password hasher for all user FQCNs returned by your user provider are the most coupling here I think, not necessarily a few classes that were installed.

So now I'm leaning more towards having a `Symfony\Component\Security\Core\Password` namespace (and keep the `Encoder` namespace deprecated). But I would love to hear any advantages of splitting that I've missed :)

chalasr commented:
Well, splitting makes sense because the moved logic is just useful on its own.
security-core don't need hashers to work, nor hashers need security-core to work.
This is enough to make the move legit :)

The fact we don't put it in a broader crypto package doesn't make it less useful as a standalone component...
Encryption is another topic and it should be in its own component as well, if we decide to have such logic in core.

We have some powerful logic for password hashing, it does deserve more visibility.

chalasr commented:
Component moved to the root namespace and renamed to `PasswordHash` / `symfony/password-hash`. 
Let me know if you still prefer `password-hashing`/`password-hasher`, I have no strong preference here.

OskarStark commented on composer.json:
```diff
@@ -74,6 +74,7 @@
         "symfony/monolog-bridge": "self.version",
         "symfony/notifier": "self.version",
         "symfony/options-resolver": "self.version",
+        "symfony/password-hash": "self.version",
```
```suggestion
        "symfony/password-hasher": "self.version",
```
WDYT?

chalasr replied:
Works for me :) Thank you!

chalasr commented:
Component renamed to `password-hasher` according to feedbacks. 

PR rebased also, now looking for votes :)

javiereguiluz commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php:
```diff
@@ -717,6 +730,117 @@ private function createEncoder(array $config)
+            ];
+        }
+
+        // run-time configured encoder
```
`encoder` -> `hasher` ?

nicolas-grekas replied:
this is in the createEncoder legacy method

javiereguiluz commented on src/Symfony/Bundle/SecurityBundle/Resources/config/schema/security-1.0.xsd:
```diff
@@ -84,6 +92,23 @@
+            <xsd:element name="migrate-from" type="xsd:string" minOccurs="0" maxOccurs="unbounded" />
+        </xsd:sequence>
+        <xsd:attribute name="class" type="xsd:string" use="required" />
+        <xsd:attribute name="algorithm" type="xsd:string" />
```
I guess there are tech reasons behind this, but it's confusing to see both `algorithm` and `hash-algorithm`

chalasr replied:
I agree with you, this feels confusing. 

The main reason is that it has been introduced for the `pbkdf2`-based encoder, which is a key derivation function that requires to configure the underlying hash algorithm to be used (which can be any of `hash_algos()` as indicated by the SecurityBundle configuration).

We might want to rename it to `pbkdf2_algorithm` or equivalent. 
However, this PR is already quite big and full of renaming, I'd like to stop doing more changes. So I'll keep this issue for another PR before 5.3 if you don't mind.

Thank you!

javiereguiluz reviewed (commented):
@chalasr thanks a lot for taking care of this!

My only minor concern is that in the code sometimes we use "hasher(s)" and other times we use "password hasher(s)" ... so it might be slightly inconsistent.

nicolas-grekas commented:
Based on https://twitter.com/TerahashCorp/status/1155129705034653698, I wonder if we shouldn't default to `bcrypt` in 5.3. That's also still PHP's default.

chalasr commented:
@nicolas-grekas I agree!  As it means revisiting NativePasswordHasher' internals and its documentation, I think it deserves its own PR (which I can do). Works for you?

nicolas-grekas commented:
Works for me!

nicolas-grekas commented on src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php:
```diff
@@ -30,6 +31,8 @@
  *
  * @final
+ *
+ * @deprecated since Symfony 5.3, use {@link UserPasswordHashCommand} instead
```
should it display a warning when run? I think we did so in the past
or is this covered by the service deprecation?

chalasr replied:
warning added

nicolas-grekas commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php:
```diff
@@ -65,6 +65,25 @@ public function getConfigTreeBuilder()
+                })
+                ->then(function ($v) {
+                    $v['password_hashers'] = array_merge($v['password_hashers'] ?? [], $v['encoders'] ?? []);
+                    $v['encoders'] = array_merge($v['password_hashers'] ?? [], $v['encoders'] ?? []);
```
shouldn't this be `$v['encoders'] = $v['password_hashers']`?
what's the point of merging `$v['encoders']` twice?

nicolas-grekas commented on src/Symfony/Bundle/SecurityBundle/Resources/config/security.php:
```diff
@@ -109,13 +113,33 @@
+        ->set('security.user_password_hasher.generic', UserPasswordHasher::class)
+            ->args([service('security.hasher_factory')])
+        ->alias('security.password_hasher', 'security.user_password_hasher.generic')
+            ->public()
```
still needed? can't we get rid of this?

chalasr replied:
now private :)

nicolas-grekas commented on src/Symfony/Bundle/SecurityBundle/Resources/config/security.php:
```diff
@@ -109,13 +113,33 @@
+
+        ->set('security.hasher_factory.generic', HasherFactory::class)
+            ->args([
+                [],
```
on one line

nicolas-grekas commented on src/Symfony/Bundle/SecurityBundle/Resources/config/console.php:
```diff
@@ -20,6 +21,16 @@
+        ->set('security.command.user_password_hash', UserPasswordHashCommand::class)
+        ->args([
+            service('security.hasher_factory'),
+            abstract_arg('hashers user classes'),
```
list of password hasher classes

nicolas-grekas commented on src/Symfony/Component/PasswordHasher/Command/UserPasswordHashCommand.php:
```diff
@@ -0,0 +1,213 @@
+            return $userClass;
+        }
+
+        if (empty($this->userClasses)) {
```
`if (!$this->userClasses) {`

nicolas-grekas commented on src/Symfony/Component/PasswordHasher/Command/UserPasswordHashCommand.php:
```diff
@@ -0,0 +1,213 @@
+        }
+
+        if (empty($this->userClasses)) {
+            throw new RuntimeException('There are no configured hashers for the "security" extension.');
```
s/hashers/password hashers

nicolas-grekas commented on src/Symfony/Component/Security/Core/Validator/Constraints/UserPasswordValidator.php:
```diff
@@ -22,12 +24,16 @@
+    private $hasherFactory;
 
-    public function __construct(TokenStorageInterface $tokenStorage, EncoderFactoryInterface $encoderFactory)
+    public function __construct(TokenStorageInterface $tokenStorage, $hasherFactory)
```
missing docblock to tell about the expected type

nicolas-grekas commented on src/Symfony/Component/Security/Http/Tests/Authenticator/HttpBasicAuthenticatorTest.php:
```diff
@@ -4,31 +4,32 @@
 
     protected function setUp(): void
     {
+
```
boo

nicolas-grekas commented on src/Symfony/Component/PasswordHasher/Hasher/HasherFactory.php:
```diff
@@ -0,0 +1,216 @@
+        }
+
+        if (null === $hasherKey) {
+            throw new \RuntimeException(sprintf('No hasher has been configured for account "%s".', \is_object($user) ? get_debug_type($user) : $user));
```
s/hasher/password hasher

nicolas-grekas commented on src/Symfony/Component/PasswordHasher/Hasher/HasherFactory.php:
```diff
@@ -0,0 +1,216 @@
+
+        if (($user instanceof HasherAwareInterface && null !== $hasherName = $user->getHasherName()) || ($user instanceof EncoderAwareInterface && null !== $hasherName = $user->getEncoderName())) {
+            if (!\array_key_exists($hasherName, $this->hashers)) {
+                throw new \RuntimeException(sprintf('The hasher "%s" was not configured.', $hasherName));
```
password hasher

nicolas-grekas commented on src/Symfony/Component/PasswordHasher/Hasher/MessageDigestPasswordHasher.php:
```diff
@@ -0,0 +1,97 @@
+
+    private function mergePasswordAndSalt(string $password, ?string $salt): string
+    {
+        if (empty($salt)) {
```
`if(!$salt)`

nicolas-grekas commented on src/Symfony/Component/PasswordHasher/Hasher/SodiumPasswordHasher.php:
```diff
@@ -0,0 +1,110 @@
+    public function __construct(int $opsLimit = null, int $memLimit = null)
+    {
+        if (!self::isSupported()) {
+            throw new LogicException('Libsodium is not available. You should either install the sodium extension, upgrade to PHP 7.2+ or use a different hasher.');
```
password hasher

nicolas-grekas commented on src/Symfony/Component/PasswordHasher/Hasher/SodiumPasswordHasher.php:
```diff
@@ -0,0 +1,110 @@
+            return \Sodium\crypto_pwhash_str($plainPassword, $this->opsLimit, $this->memLimit);
+        }
+
+        throw new LogicException('Libsodium is not available. You should either install the sodium extension, upgrade to PHP 7.2+ or use a different hasher.');
```
password hasher

nicolas-grekas commented on src/Symfony/Component/PasswordHasher/Hasher/SodiumPasswordHasher.php:
```diff
@@ -0,0 +1,110 @@
+            return \Sodium\crypto_pwhash_str_needs_rehash($hashedPassword, $this->opsLimit, $this->memLimit);
+        }
+
+        throw new LogicException('Libsodium is not available. You should either install the sodium extension, upgrade to PHP 7.2+ or use a different hasher.');
```
password hasher

nicolas-grekas commented on src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php:
```diff
@@ -59,14 +67,29 @@ protected function checkAuthentication(UserInterface $user, UsernamePasswordToke
 
-            if ($this->userProvider instanceof PasswordUpgraderInterface && method_exists($encoder, 'needsRehash') && $encoder->needsRehash($user->getPassword())) {
-                $this->userProvider->upgradePassword($user, $encoder->encodePassword($presentedPassword, $user->getSalt()));
+            if ($this->userProvider instanceof PasswordUpgraderInterface && method_exists($hasher, 'needsRehash') && $hasher->needsRehash($user->getPassword())) {
```
should we find a way to get rid of the method_exists check?

wouterj replied:
I think we can already get rid of it, as `needsRehash()` is part of `PasswordHasherInterface`.

nicolas-grekas commented on src/Symfony/Component/Security/Core/Tests/Encoder/TestPasswordEncoderInterface.php:
```diff
@@ -13,6 +13,10 @@
 use Symfony\Component\Security\Core\Encoder\PasswordEncoderInterface;
 
+/**
+ * @package Symfony\Component\Security\Core\Tests\Encoder
```
needed?

nicolas-grekas commented on src/Symfony/Component/Security/Http/composer.json:
```diff
@@ -18,14 +18,15 @@
     },
     "require-dev": {
         "symfony/cache": "^4.4|^5.0",
+        "symfony/password-hasher": "^5.3",
```
not needed

chalasr replied:
needed it is, password-hasher interfaces are used in tests

nicolas-grekas replied:
but it comes as a mandatory transitive dep, isn't it?

chalasr replied:
indeed, sorry! removed

nicolas-grekas reviewed (commented):
LGTM thanks. Here are some nitpicking and a few questions.

nicolas-grekas commented:
Oh and please add a few lines in changelog / upgrade files (and add one in the new component)

wouterj commented on src/Symfony/Component/PasswordHasher/README.md:
```diff
@@ -0,0 +1,14 @@
+
+The PasswordHasher component provides password hashing utilities for both old and modern algorithms,
+along with hash migration capabilities to keep applications on top of security standards.
+
```
Please add a "Getting Started" section like https://github.com/symfony/translation#getting-started (I'll submit a PR for the components that don't have this yet).

wouterj commented on src/Symfony/Component/PasswordHasher/composer.json:
```diff
@@ -0,0 +1,33 @@
+{
+    "name": "symfony/password-hasher",
+    "type": "library",
+    "description": "Symfony Password Hashing Component",
```
```suggestion
    "description": "Provides password hashing utilities for both old and modern algorithms, along with hash migration capabilities",
```

(Symfony recently changed all its `description`s to be more descriptive)

wouterj commented on src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasher.php:
```diff
@@ -0,0 +1,64 @@
+use Symfony\Component\Security\Core\User\UserInterface;
+
+/**
+ * A generic password hasher.
```
```suggestion
 * Hashes passwords based on the user and the HasherFactory.
```

"generic" doesn't say much :)

wouterj commented on src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasherInterface.php:
```diff
@@ -0,0 +1,34 @@
+     */
+    public function hashPassword(UserInterface $user, string $plainPassword): string;
+
+    public function isPasswordValid(UserInterface $user, string $raw): bool;
```
```suggestion
    public function isPasswordValid(UserInterface $user, string $plainPassword): bool;
```

I think it's good to use only "plaintext"/"plain password" to be clear on what is expected.

wouterj replied:
Maybe also add a PHPdoc here, e.g: "Checks if the plaintext password matches the user's password"

wouterj commented on src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasherInterface.php:
```diff
@@ -0,0 +1,34 @@
+interface UserPasswordHasherInterface
+{
+    /**
+     * Hashes the plain password.
```
```suggestion
     * Hashes the plain password based on the user.
```

(maybe explicitly explain that this does not *set* the password on the user? not sure)

wouterj commented on src/Symfony/Component/PasswordHasher/Hasher/HasherFactory.php:
```diff
@@ -0,0 +1,216 @@
+/**
+ * A generic hasher factory implementation.
+ *
+ * @author Nicolas Grekas <p@tchork.com>
```
```suggestion
 * @author Nicolas Grekas <p@tchwork.com>
```

wouterj commented on src/Symfony/Component/PasswordHasher/Exception/LogicException.php:
```diff
@@ -0,0 +1,19 @@
+/**
+ * @author Robin Chalas <robin.chalas@gmail.com>
+*/
+class LogicException extends \LogicException implements ExceptionInterface
```
The component uses a mixture of `\LogicException` and this logic exception. Personally, I'm always using `\LogicException` as I believe only runtime exceptions should be catched (and thus need to be identified using the custom `ExceptionInterface`).

Whatever we choose, we should use it consistently in the component :)

wouterj commented on src/Symfony/Component/PasswordHasher/Hasher/MigratingPasswordHasher.php:
```diff
@@ -0,0 +1,65 @@
+ * Hashes passwords using the best available hasher.
+ * Verifies them using a chain of hashers.
+ *
+ * /!\ Don't put a PlaintextPasswordHasher in the list as that'd mean a leaked hash
+ * could be used to authenticate successfully without knowing the cleartext password.
```
Should we make this a hard requirement by checking for `PlaintextPasswordHasher`?

chalasr replied:
It was decided to not do that when the equivalent encoder was introduced, see https://github.com/symfony/symfony/pull/31597#discussion_r286942338

wouterj commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php:
```diff
@@ -65,6 +65,25 @@ public function getConfigTreeBuilder()
             ->end()
+            ->beforeNormalization()
+                ->ifTrue(function ($v) {
+                    if (($v['encoders'] ?? false)) {
```
```suggestion
                    if ($v['encoders'] ?? false) {
```

wouterj commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php:
```diff
@@ -65,6 +65,25 @@ public function getConfigTreeBuilder()
+                        return true;
+                    }
+
+                    if ($v['password_hashers'] ?? false) {
+                        return true;
+                    }
```
```suggestion
                    return isset($v['password_hashers']);
```

wouterj commented on src/Symfony/Component/Security/Http/EventListener/PasswordMigratingListener.php:
```diff
@@ -72,7 +82,11 @@ public function onLoginSuccess(LoginSuccessEvent $event): void
+        try {
+            $passwordUpgrader->upgradePassword($user, $passwordHasher->hash($plaintextPassword, $user->getSalt()));
+        } catch (InvalidPasswordException $e) {
+            throw new BadCredentialsException('Bad credentials.', 0, $e);
```
This is already catched by `CheckCredentialsListener` (failed logins never result in password migration).

chalasr replied:
nice catch, fixed

wouterj reviewed (changes requested):
:+1: Looks good! Most of my comments are minor doc related suggestions.

chalasr commented:
@javiereguiluz @nicolas-grekas @wouterj Thank you very much for the reviews. 
All comments have been addressed

nicolas-grekas commented on src/Symfony/Bundle/SecurityBundle/CHANGELOG.md:
```diff
@@ -1,6 +1,16 @@
+5.3
+---
+
+ * Deprecate `UserPasswordEncoderCommand` class and the corresponding `user:encode-password` command,
```
final dots should be removed, according to the clarified policy

nicolas-grekas commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php:
```diff
@@ -717,6 +731,117 @@ private function createEncoder(array $config)
+                $config['algorithm'] = 'native';
+                $config['native_algorithm'] = \PASSWORD_ARGON2I;
+            } else {
+                throw new InvalidConfigurationException(sprintf('Algorithm "argon2i" is not available. Either use "%s" or upgrade to PHP 7.2+ instead.', \defined('SODIUM_CRYPTO_PWHASH_ALG_ARGON2ID13') ? 'argon2id", "auto' : 'auto'));
```
PHP 7.2 it is already :)
can't we remove the defined() check above?

chalasr replied:
All "upgrade to PHP 7.2+" advices removed from the new code. I'll review the legacy one on 5.2.
I may be wrong as I can't remember exactly, but I believe we can't be 100% sure this constant will be defined for PHP versions prior to 7.3 (too much differences in sodium integrations). I'd keep it as-is for now :)

nicolas-grekas reviewed (approved):
Thanks! Good to me, ~once UPGRADE files are updated~ :)

wouterj commented:
Thanks a lot @chalasr! I can't wait to see all the deprecated stuff be removed in 6.0 :)
